lib/wibox/layout/grid.lua

```diff
@@ -0,0 +1,484 @@
+
+        if idx and l and l.remove then
+            l:remove(idx, false)
+        else
+            ret = false
+        end
+    end
+
+    return #args > (recursive and 1 or 0) and ret
+end
+
+function grid:get_children()
+    return self._private.children
+end
+
+function grid:set_children(children)
+    self:reset()
+    if #children > 0 then
+        self:add(unpack(children))
+    end
+end
+
+--- Replace the first instance of `widget` in the layout with `widget2`
+-- @param widget The widget to replace
+-- @param widget2 The widget to replace `widget` with
+-- @tparam[opt=false] boolean recursive Digg in all compatible layouts to find the widget.
+-- @treturn boolean If the operation is successful
+function grid:replace_widget(widget, widget2, recursive)
+    local idx, l = self:index(widget, recursive)
+
+    if idx and l then
+        l:set(idx, widget2)
+        return true
+    end
+
+    return false
+end
+
+function grid:swap_widgets(widget1, widget2, recursive)
+    base.check_widget(widget1)
+    base.check_widget(widget2)
+
+    local idx1, l1 = self:index(widget1, recursive)
+    local idx2, l2 = self:index(widget2, recursive)
+
+    if idx1 and l1 and idx2 and l2 and (l1.set or l1.set_widget) and (l2.set or l2.set_widget) then
+        if l1.set then
+            l1:set(idx1, widget2)
+        elseif l1.set_widget then
+            l1:set_widget(widget2)
+        end
+        if l2.set then
+            l2:set(idx2, widget1)
+        elseif l2.set_widget then
+            l2:set_widget(widget1)
+        end
+
+        return true
+    end
+
+    return false
+end
+
+function grid:set(index, widget2)
+    if (not widget2) or (not self._private.rows[index]) then return false end
+
+    base.check_widget(widget2)
+
+    self._private.rows[index] = widget2
+
+    self:emit_signal("widget::layout_changed")
+
+    return true
+end
+
+--- Insert a new widget in the layout at position `index`
+-- @tparam number index The position
+-- @param widget The widget
+-- @treturn boolean If the operation is successful
+function grid:insert(index, widget)
+    local TODO_TOTAL = math.huge --FIXME
+    if not index or index < 1 or index > TODO_TOTAL then return false end
+
+    base.check_widget(widget)
+    table.insert(self._private.rows, index, widget) --FIXME
+    self:emit_signal("widget::layout_changed")
+
+    return true
+end
+
+-- Fit the grid layout into the given space
+-- @param context The context in which we are fit.
+-- @param orig_width The available width.
+-- @param orig_height The available height.
+function grid:fit(context, orig_width, orig_height)
+    local max_col, max_row = get_cell_sizes(self, context, orig_width, orig_height)
+
+    -- Now that all fit is done, get the maximum
+    local used_max_h, used_max_w = 0, 0
+
+    for row_idx = 1, #max_row do
+        -- The other widgets will be discarded
+        if used_max_h + max_row[row_idx] > orig_height then
+            break
+        end
```

Comment:
I don't think that the other widgets will be discarded. Instead, I read the code so that they will be "forced into the available space", won't they?

In other words, I think that this `if` and the `break` can be removed without problems (same for the next for loop; no it is "not really" a problem if `:fit` returns numbers that are larger than the available space, `base.fit_widget` will clean that up and force the result into the available space).